honeybadgermpc/progs/fixedpoint.py

```diff
@@ -0,0 +1,303 @@
+"""
+
+
+async def div2m(ctx, x, k, m):
+    r1, r1_bits = await random2m(ctx, m)
+    r2, _ = await random2m(ctx, k + KAPPA - m)
+    r2 = ctx.Share(r2.v * Field(2) ** m)
+
+    c = await (x + r2 + r1 + Field(2) ** (k - 1)).open()
+    c2 = int(c) % (2 ** m)
+    u = await bit_ltl(ctx, c2, r1_bits)
+    a2 = ctx.Share(Field(c2) - r1.v + (2 ** m) * u.v)
+    return a2
+
+
+"""
+Given the secret shared [x] calcuate the secret shares of [x%2^m] for known  public m.
+This is calcuated by first calculating the value of [x//2^m] usiinig div2m and substracting that from [x]
+
+Returns the value: [x % 2^m]
+"""
+
+
+async def trunc(ctx, x, k, m):
+    a2 = await div2m(ctx, x, k, m)
+    d = ctx.Share((x.v - a2.v) / (Field(2)) ** m)
+    return d
+
+
+class FixedPoint(object):
+    def __init__(self, ctx, x):
+        self.ctx = ctx
+        if type(x) in [int, float]:
+            self.share = ctx.preproc.get_zero(ctx) + ctx.Share(int(x * 2 ** F))
+        elif type(x) is ctx.Share:
+            self.share = x
+        else:
+            raise NotImplementedError
+
+    def __add__(self, x):
+        if type(x) is FixedPoint:
+            return FixedPoint(self.ctx, self.share + x.share)
+
+    def __sub__(self, x):
+        if type(x) is FixedPoint:
+            return FixedPoint(self.ctx, self.share - x.share)
+        raise NotImplementedError
+
+    """
+    TODO: replacing * in __mul__ in a await statement does not work.
+
+    Multiplty the two number using normal multiplication of numbers.
+    Truncate the last F bits of resulting 2*K bit number.
+    """
+
+    async def __mul__(self, x):
+        if type(x) is FixedPoint:
+            start_time = time.time()
+            res_share = await (self.share * x.share)
+            end_time = time.time()
+            logging.info("Multiplication time: %.2f", end_time - start_time)
+            start_time = time.time()
+            res_share = await trunc_pr(self.ctx, res_share, 2 * K, F)
+            end_time = time.time()
+            logging.info("Trunc time: %.2f", end_time - start_time)
+            return FixedPoint(self.ctx, res_share)
+        raise NotImplementedError
+
+    async def open(self):
+        x = (await self.share.open()).value
+        if x >= 2 ** (K - 1):
+            x = -(p - x)
+        return float(x) / 2 ** F
+
+    def neg(self):
+        return FixedPoint(self.ctx, Field(-1) * self.share)
+
+    """
+    Compute the last K-1 bits and check if the resulting number is 0 or 1.
+    """
+
+    async def ltz(self):
+        t = await trunc(self.ctx, self.share, K, K - 1)
+        return self.ctx.Share(-t.v)
+
+    """
+    Check whether the self is than is x.
+    """
+
+    async def lt(self, x):
+        return await (self - x).ltz()
+
+    async def div(self, x):
+        if type(x) in [float, int]:
+            return await self.__mul__(FixedPoint(self.ctx, 1.0 / x))
+        raise NotImplementedError
```

Comment:
### Comments for the `FixedPoint` class.

Classes in Python 3 extend `object` by default, so you can omit it:

```python
class FixedPoint:
```

To check the type of an argument, `isinstance()` is the preferred way:

```python
 if isinstance(x, (float, int)):
```

#### About dispatching based on the type of an argument
Just as a note, since the behaviour of some methods depends on the type you may find [`functools.singledispatchmethod`](https://docs.python.org/3/library/functools.html#functools.singledispatchmethod) interesting. In order to use it we would need to upgrade to Python 3.8 though, or use an approach inspired by https://stackoverflow.com/a/24064102/2769475.